### PR TITLE
Gradient circuits 4/n: update LinearCombination

### PR DESCRIPTION
**Pull request by zaqqwerty**
For after #532.  During implementation of that PR, I discovered a more efficient way to represent the gradient structure: separating the old `batch_mapper` into a weights tensor and an index tensor.  In the updated tests in this PR, it can be seen that this eliminates the zero padding previously required.

### CODE REVIEW

**Comment by zaqqwerty** [comment]
For some reason the tests do not run and instantly fail - the summary says "GitHub Actions has encountered an internal error when running your job."

**Comment by jaeyoo** [comment]
> For some reason the tests do not run and instantly fail - the summary says "GitHub Actions has encountered an internal error when running your job."

it will disappear after several hours. no worries, please take your time :)

File: tensorflow_quantum/python/differentiators/linear_combination.py [hunk 1]
```diff
@@ -111,7 +111,6 @@ def __init__(self, weights, perturbations):
         self.n_non_zero_perturbations = tf.gather(
             tf.shape(self.non_zero_perturbations), 0)
 
-    @tf.function
```
Comment (MichaelBroughton) [inline tensorflow_quantum/python/differentiators/linear_combination.py:114]: Wouldn't we want to keep this around ?

**Review by MichaelBroughton — COMMENTED** [review]
Changes look good. Just one concern on removing the decorator.

File: tensorflow_quantum/python/differentiators/linear_combination.py (hunk 1 above)
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/linear_combination.py]: Added back, I had removed during debugging and forgot

**Review by MichaelBroughton — APPROVED** [review]

**Pull request merged**